File: lib/aerox/scene/components/SceneComponent.cpp
#include <aerox/scene/components/SceneComponent.hpp>
#include "aerox/scene/objects/SceneObject.hpp"

namespace aerox::scene {

math::Vec3<> SceneComponent::GetRelativeLocation() const {
  return GetWorldTransform().location;
}

math::Quat SceneComponent::GetRelativeRotation() const {
  return GetWorldTransform().rotation;
}

math::Vec3<> SceneComponent::GetRelativeScale() const {
  return GetRelativeTransform().scale;
}

math::Vec3<> SceneComponent::GetWorldLocation() const {
  return GetWorldTransform().location;
}

math::Quat SceneComponent::GetWorldRotation() const {
  return GetWorldTransform().rotation;
}

math::Vec3<> SceneComponent::GetWorldScale() const {
  return GetWorldTransform().scale;
}

math::Transform SceneComponent::GetWorldTransform() const {
  if(const auto parent = GetParent()) {
    return GetRelativeTransform().RelativeTo(parent->GetWorldTransform());
  }

  return GetRelativeTransform();
  //return GetRelativeTransform();
}

void SceneComponent::SetRelativeLocation(const math::Vec3<> &val) {
  SetRelativeTransform(val,{},{});
}

void SceneComponent::SetRelativeRotation(const math::Quat &val) {
  SetRelativeTransform({},val,{});
}

void SceneComponent::SetRelativeScale(const math::Vec3<> &val) {
  SetRelativeTransform({},{},val);
}

void SceneComponent::SetWorldLocation(const math::Vec3<> &val) {
  SetWorldTransform(val,{},{});
}

void SceneComponent::SetWorldRotation(const math::Quat &val) {
  SetWorldTransform({},val,{});
}

void SceneComponent::SetWorldScale(const math::Vec3<> &val) {
  
  SetWorldTransform({},{},val);
}

void SceneComponent::SetWorldTransform(const math::Transform &val) {
  // _relativeTransform = val;
  if(GetParent()) {
    const auto targetWorldMatrix = val.Matrix();
    const auto thisWorldMatrix = GetWorldTransform().Matrix();

    const auto thisToTarget = inverse(thisWorldMatrix) * targetWorldMatrix;
    const auto thisToParent = GetRelativeTransform().Matrix();
    
    SetRelativeTransform(thisToTarget * inverse(thisToParent));
    return;
  }

  SetRelativeTransform(val);
}

void SceneComponent::SetWorldTransform(
    const std::optional<math::Vec3<>> &location,
    const std::optional<math::Quat> &rotation,
    const std::optional<math::Vec3<>> &scale) {
  auto existingTransform = GetWorldTransform();
  SetWorldTransform({location.value_or(existingTransform.location),rotation.value_or(existingTransform.rotation),scale.value_or(existingTransform.scale)});
}

math::Transform SceneComponent::GetRelativeTransform() const {
  return _relativeTransform;
}

void SceneComponent::SetRelativeTransform(const math::Transform &val) {
  _relativeTransform = val;
}

void SceneComponent::SetRelativeTransform(
    const std::optional<math::Vec3<>> &location,
    const std::optional<math::Quat> &rotation,
    const std::optional<math::Vec3<>> &scale) {
  auto existingTransform = GetWorldTransform();
  SetRelativeTransform({location.value_or(existingTransform.location),rotation.value_or(existingTransform.rotation),scale.value_or(existingTransform.scale)});
  
}

SceneComponent * SceneComponent::GetParent() const {
  return _parent;
}

void SceneComponent::AttachTo(const std::weak_ptr<SceneComponent> &parent) {
  _parent = parent.lock().get();
}
}

File: aerox.Runtime.Scene/Components/SceneComponent.cs
using aerox.Runtime.Math;
using aerox.Runtime.Scene.Graphics;

namespace aerox.Runtime.Scene.Components;

public class SceneComponent : Component, ISceneDrawable
{
    private Transform _relativeTransform = new();
    
    public SceneComponent? Parent { get; private set; }

    public readonly List<SceneComponent> Children = []; 

    public void AttachTo(SceneComponent component)
    {
        Parent = component;
        Parent.Children.Add(this);
    }
    
    public Transform GetRelativeTransform()
    {
        return _relativeTransform.Clone();
    }
    
    public Vector3<float> GetRelativeLocation()
    {
        return _relativeTransform.Location.Clone();
    }
    
    public Quaternion GetRelativeRotation()
    {
        return _relativeTransform.Rotation.Clone();
    }
    
    public Vector3<float> GetRelativeScale()
    {
        return _relativeTransform.Scale.Clone();
    }
    
    public void SetRelativeTransform(Transform transform)
    {
        _relativeTransform = transform.Clone();
    }
    
    public void SetRelativeLocation(Vector3<float> location)
    {
        _relativeTransform.Location = location.Clone();
    }
    
    public void SetRelativeRotation(Quaternion rotation)
    {
        _relativeTransform.Rotation = rotation.Clone();
    }
    
    public void SetRelativeScale(Vector3<float> scale)
    {
        _relativeTransform.Scale = scale.Clone();
    }

    public Transform GetWorldTransform()
    {
        if (Parent == null) return GetRelativeTransform();

        Matrix4 parentTransform = Parent.GetWorldTransform();
        return parentTransform * _relativeTransform;
    }

    public void SetWorldTransform(Transform worldTransform)
    {
        if (Parent != null)
        {
            Matrix4 targetWorldMatrix = worldTransform;
            Matrix4 thisWorldMatrix = GetWorldTransform();

            var thisToTarget = thisWorldMatrix.Inverse() * targetWorldMatrix;
            Matrix4 thisToParent = GetRelativeTransform();

            SetRelativeTransform(thisToTarget * thisToParent.Inverse());
            return;
        }
        
        SetRelativeTransform(worldTransform);
    }
    
    protected virtual void CollectSelf(SceneFrame frame, Matrix4 parentTransform,Matrix4 myTransform)
    {
    }

    protected virtual void CollectChildren(SceneFrame frame, Matrix4 parentTransform,Matrix4 myTransform)
    {
        for (var i = 0; i < Children.Count; i++)
        {
            Children[i].Collect(frame,myTransform);
        }
    }

    public virtual void Collect(SceneFrame frame, Matrix4 parentSpace)
    {
        var myTransform = parentSpace * GetRelativeTransform();
        CollectSelf(frame,parentSpace,myTransform);
        CollectChildren(frame,parentSpace,myTransform);
    }
}